### ISSUE TITLE
Multiple notifications being triggered

### ISSUE THREAD

**Issue by shabbir-taulia** [argoproj-labs/argocd-notifications#90]
Hello,

I am using the latest version of argocd-notifications and seeing strange behavior with the way notifications are being sent. For just one successful app sync there are multiple notifications being triggered for example - our current workflow is:
1. PR Merged in K8S repo
2. Argo CD app sync (automated sync)
3. Send Slack alert
4. Trigger tests via Jenkins 

The last step now is being constantly triggered - controller logs
```
time="2020-05-13T05:01:27Z" level=info msg="Start processing" app=argocd/test-app
time="2020-05-13T05:01:27Z" level=info msg="Trigger jenkins result: true" app=argocd/test-app
time="2020-05-13T05:01:27Z" level=info msg="Sending jenkins notification" app=argocd/test-app
time="2020-05-13T05:01:27Z" level=info msg="Sending jenkins notification" app=argocd/test-app
time="2020-05-13T05:01:27Z" level=info msg="Sending on-sync-succeeded notification" app=argocd/test-app
time="2020-05-13T05:01:28Z" level=info msg="Trigger jenkins result: true" app=argocd/test-app
time="2020-05-13T05:01:28Z" level=info msg="Sending jenkins notification" app=argocd/test-app
time="2020-05-13T05:01:28Z" level=info msg="Sending jenkins notification" app=argocd/test-app
```

The configmap:
```
    triggers:
      - name: on-sync-succeeded
        condition: app.status.operationState.phase == 'Succeeded' and app.status.health.status == 'Healthy'
        enabled: true
      - name: jenkins
        condition: app.status.operationState.phase == 'Succeeded' and app.status.health.status == 'Healthy'
        enabled: true
        template: jenkins
      - name: on-sync-failed
        enabled: true
      - name: on-sync-running
        condition: app.status.health.status == 'Progressing'
        enabled: true
    subscriptions:
    - recipients:
      - webhook:jenkins
      trigger: jenkins
    templates:
    - name: jenkins
      webhook:
        jenkins:
          method: POST
```

Maybe my configuration is not correct but currently this is causing multiple tests being triggered in Jenkins at the same time - any help or guidance is appreciated. Thank you!

**Comment by julientard** [comment]
Hello, 

I think I have the same issue.
If I understand well, all triggers are played against every recipients, whatever trigger is chosen in the subscriptions object.

I think the field `trigger` inside `subscriptions` is just ignored.

Here is my configuration :
```
triggers:
- name: on-sync-failed
  enabled: true
- name: on-sync-succeeded
  enabled: true
subscriptions:
- recipients:
  - slack:slack-channel-1
  trigger: on-sync-failed
- recipients:
  - slack:slack-channel-2
  trigger: on-sync-succeeded
```

In this situation both notifications are sent on both slack channels.

**Comment by julientard** [comment]
Ok it seems the field `trigger` does not work. Instead it works for me with the field `triggers` which is a list.

**Comment by YoshiChladny** [comment]
I'm running into the same issue. When adding webhooks to the global `subscriptions:` alongside slack, if any trigger is met, all of the notifications get sent to every recipient.

> Ok it seems the field trigger does not work. Instead it works for me with the field triggers which is a list.
@julientard do you have an example of this config? From the examples in the docs, setting triggers per notification

defining triggers per subscription isn't notifying the subscriptions: is there something missing here?
```
data:
  config.yaml: |
    subscriptions:
    - recipients:
      - slack:argo-notifications
      triggers: 
        - on-sync-on-sync-succeeded 
    - recipients:
      - slack:argo-notifications2
      triggers:
      - on-sync-failed
```

Also, AFAIK there is not a way to define a `condition:` within a non-global subscription (this may be a separate issue). 

the following will fail to be parsed, but works as a global subscription:
```
    - recipients:
      - webhook:myhook
      triggers:
        - name: sync-operation-succeeded
          condition: app.status.operationState.phase in ['Succeeded']
          template: sync-operation-status-change
    templates:
    - name: sync-operation-status-change
    webhook:
      myhook:
        method: POST
        path: /webhooks/myendpoint
        body: |
            { 
              {{if eq .app.status.operationState.phase "Succeeded"}} 
              "content":
              "Sync for {{.app.metadata.name}}:Sync Succeeded. Sync operation details are available - {{app.context.argocdUrl}}/applications/{{.app.metadata.name}}"
              {{end}}
            }
```

I may be missing something here, but any clarity would be greatly appreciated.

**Comment by julientard** [comment]
Hello @YoshiChladny , 

I see several things here.
On your first issue, you did not enabled your triggers before.
Default triggers need first to be enabled like this :

```
triggers:
- name: on-sync-failed
  enabled: true
- name: on-sync-succeeded
  enabled: true
```

Also on-sync-on-sync-succeeded is not a default trigger, and the indentation is wrong on this line, but i guess it's just a typo.

About your second issue, I would first declare the trigger, and then call it by it's name:
```
triggers:
- name: sync-operation-succeeded
  condition: app.status.operationState.phase in ['Succeeded']
  template: sync-operation-status-change
  enabled: true # not sure if this is necessary for custom trigger
```
Then
```
subscriptions:
- recipients:
  - webhook:myhook
  triggers:
  - sync-operation-succeeded
```
Something like this.

**Comment by YoshiChladny** [comment]
Thanks @julientard I think the issue may stem from using both built in triggers for slack, alongside custom webhooks.

if I add both webhook, and slack notifications to my app, slack gets constantly spammed with notifications, as @shabbir-taulia mentioned.

`kubectl patch app <my-app> -n argocd -p '{"metadata": {"annotations": {"recipients.argocd-notifications.argoproj.io":"webhook:myhook, slack:argo-notifications"}}}' --type merge`

 configmap:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: argocd-notifications-cm
data:
  config.yaml: |
    triggers:
    ## custom triggers:
    - name: sync-operation-succeeded
      condition: condition: app.status.operationState.phase == 'Succeeded' and app.status.health.status == 'Healthy'
      template: sync-operation-status-change  
    ## slack triggers:
    - name: on-sync-succeeded
      enabled: true
    subscriptions:
    - recipients:
      - webhook:myhook
      triggers:
      - sync-operation-succeeded
    - recipients:
      - slack:argo-notifications
      triggers: 
      - on-sync-succeeded
    templates:
    - name: sync-operation-status-change
      webhook:
        myhook:
          method: POST
          path: /webhooks/mypath
          body: |
              { 
                {{if eq .app.status.operationState.phase "Succeeded"}} 
                "content":
                "Sync for {{.app.metadata.name}}: Succeeded ✓. Sync operation details are available at {{.context.argocdUrl}}/applications/{{.app.metadata.name}}"
                {{end}}
              }
```

However adding ONLY a single recipient to the app, and custom hooks, or the slack notifications will work as intended.

`kubectl patch app <my-app> -n argocd -p '{"metadata": {"annotations": {"recipients.argocd-notifications.argoproj.io":"<webhook:myhook OR slack:argo-nofications>"}}}' --type merge`

it doesn't seem right that these would be mutually exclusive.

**Comment by eakman-nomad** [comment]
@shabbir-taulia, @julientard, @YoshiChladny did you guys ever find a solution or workaround for this by any chance? I think I'm running into the same issue. My secret and configmap looks as follows:
```
---
apiVersion: v1
kind: Secret
metadata:
  name: argocd-notifications-secret
stringData:
  notifiers.yaml: |
    webhook:
    - name: gh-deployment-wh
      url: http://github-deployments-svc.deployments-manager-ns.svc/api/deployments/
      headers:
      - name: Content-Type
        value: application/json
type: Opaque
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: argocd-notifications-cm
data:
  config.yaml: |
    triggers:
      - name: sync-running
        condition: app.status.operationState.phase in ['Running']
        template: gh-deployment-create-template
        enabled: true
      - name: sync-failed
        condition: app.status.operationState.phase in ['Error', 'Failed']
        template: gh-deployment-failure-template
      - name: sync-succeeded
        condition: app.status.operationState.phase in ['Succeeded']
        template: gh-deployment-success-template

    templates:
      - name: gh-deployment-create-template
        webhook:
          gh-deployment-wh:
            method: POST
            body: |
              {
                "repo_name": "{{call .repo.FullNameByRepoURL .app.spec.source.repoURL}}",
                "environment": "{{ .app.spec.destination.namespace }}",
                "ref": "{{ .app.spec.source.targetRevision }}"
              }

      - name: gh-deployment-failure-template
        webhook:
          gh-deployment-wh:
            method: PATCH
            body: |
              {
                "repo_name": "{{call .repo.FullNameByRepoURL .app.spec.source.repoURL}}",
                "environment": "{{ .app.spec.destination.namespace }}",
                "ref": "{{ .app.spec.source.targetRevision }}",
                "state": "failure"
              }

      - name: gh-deployment-success-template
        webhook:
          gh-deployment-wh:
            method: PATCH
            body: |
              {
                "repo_name": "{{call .repo.FullNameByRepoURL .app.spec.source.repoURL}}",
                "environment": "{{ .app.spec.destination.namespace }}",
                "ref": "{{ .app.spec.source.targetRevision }}",
                "state": "success"
              }
```

If I specify just my custom webhook in the recipients annotation on my application (e.g. `recipients.argocd-notifications.argoproj.io: webhook:gh-deployment-wh`), it works exactly as expected and triggers the webhook once. But if I specify slack in addition to my custom webhook `recipients.argocd-notifications.argoproj.io: webhook:gh-deployment-wh, slack:ship`, my webhook gets triggered multiple times.

**Comment by eakman-nomad** [comment]
FYI, what I ended up doing is setting my 'subscriptions` in my globally in my notifications `ConfigMap` as follows:

```
    subscriptions:
    # global subscription restricted to applications with matching labels only
    - recipients:
      - webhook:gh-deployment-wh
      triggers:
        - sync-running
        - sync-failed
        - sync-succeeded
      selector: notify.acme.com/gh-deployments=1

    - recipients:
      - slack:ship
      triggers:
        - on-sync-succeeded
        - on-sync-failed
        - on-health-degraded
        - on-sync-running
        - on-sync-status-unknown
      selector: notify.acme.com/ship=1
```

I specified the triggers for each recipient and instead of using the `recipients.argocd-notifications.argoproj.io: webhook:<webhook-name>` annotation syntax, I just use selectors e.g.:

```
...
metadata:
  name: my-app
  labels:
    notify.nomadhealth.com/gh-deployments: '1'
...
```

**Comment by booleanbetrayal** [comment]
We are seeing the same behavior. When a single recipient is set on either the `Application` annotation, or in the `argocd-notifications-cm`'s `config.yaml`, everything works fine. The moment we specify multiple notification channels in the annotation or, in the `recipients` block of of the config file, then we start getting spammed again. Likewise, I'm seeing the spamming occur when I have isolated notification channels defined in multiple `recipient` blocks that hit an identical `selector`.

**Comment by booleanbetrayal** [comment]
Turns out that it is likely that this was related to https://github.com/argoproj-labs/argocd-notifications/issues/79 in conjunction with not having a slack bot added to the appropriate channel. Hopefully we can get a build with the latest fixes cut soon.

**Comment by kumarprd** [comment]
How to control the frequency of sending notifications ? It keeps on sending notification on condition match. It should send notification only once when state is changed.

**Comment by alexmt** [comment]
@kumarprd , 

Triggers now have `oncePer` property that might help. The notification is sent once per value of the field specified by `oncePer`. For example following trigger notifies only once per commit sha:

```yaml
  trigger.on-deployed: |
    - description: Application is synced and healthy. Triggered once per commit.
      oncePer: app.status.sync.revision
      send:
      - app-deployed
      when: app.status.operationState.phase in ['Succeeded'] and app.status.health.status == 'Healthy'
```

**Comment by tsunamishaun** [comment]
I've gone through and added the `oncePer` to my triggers and it doesn't seem to help. My use-case and configuration is almost identical to @YoshiChladny 's, seeing slack fire twice for the same trigger/id and I use both webhooks and slack in my config. I've compared the status, phase, health and ids for each duplicate event and they are indeed identical. Not sure what to do?

**Comment by aslamkhan-dremio** [comment]
I am seeing same issue even after adding oncePer to triggers

**Comment by hendrikmaus** [comment]
> @kumarprd ,
> 
> Triggers now have `oncePer` property that might help. The notification is sent once per value of the field specified by `oncePer`. For example following trigger notifies only once per commit sha:
> 
> ```yaml
>   trigger.on-deployed: |
>     - description: Application is synced and healthy. Triggered once per commit.
>       oncePer: app.status.sync.revision
>       send:
>       - app-deployed
>       when: app.status.operationState.phase in ['Succeeded'] and app.status.health.status == 'Healthy'
> ```

I use this configuration exactly and often get a duplicated message of the "on-deployed" trigger. I am only using Slack as a receiver, without the bot functionality.

Considering the default triggers being set to:

```yaml
defaultTriggers: |
  - on-deployed
  - on-sync-running
```

Expected notifications in order:

- on-sync-running
- on-deployed

Actual notifications sent:

- on-deployed
- on-sync-running
- on-deployed

Though this does not happen every time. I haven't yet figured out which edge case causes the duplicated "on-deployed" message.

*As a hunge, I re-ordered the triggers in the `defaultTriggers` config, so that "on-deployed" is last in the list.*

**Comment by AbdulRahmanAlHamali** [comment]
This issue is happening intermittently for us as well, we're using v1.1.1

**Comment by hendrikmaus** [comment]
Good that you mention it; I run 1.2.1
